src/factory.rs

```diff
@@ -543,6 +546,281 @@ impl Factory {
+                occlusion_strength: mat.occlusion_texture().map_or(1.0, |t| t.strength()),
+                emissive_factor: mat.emissive_factor(),
+                normal_scale: mat.normal_texture().map_or(1.0, |t| t.scale()),
+                base_color_map,
+                normal_map,
+                emissive_map,
+                metallic_roughness_map,
+                occlusion_map,
+            }
+        }
+    }
+
+    /// Loads a `glTF` mesh.
+    ///
+    /// Note that multiple meshes are returned, since the concept of a mesh is
+    /// different in `glTF` to `three`. A glTF mesh consists of one or more
+    /// *primitives*, which are equivalent to `three` meshes.
+    fn load_gltf_mesh(
+        &mut self,
+        mesh: &gltf::Mesh,
+        buffers: &gltf_importer::Buffers,
+        base: &Path,
+    ) -> Vec<Mesh> {
+        let mut primitives = Vec::new();
+        for primitive in mesh.primitives() {
+            use gltf_utils::PrimitiveIterators;
+            let mut faces = vec![];
+            if let Some(mut iter) = primitive.indices_u32(buffers) {
+                while let (Some(a), Some(b), Some(c)) = (iter.next(), iter.next(), iter.next()) {
+                    faces.push([a, b, c]);
+                }
+            }
+            let vertices = primitive
+                .positions(buffers)
+                .unwrap()
+                .map(|x| x.into())
+                .collect();
+            let normals = if let Some(iter) = primitive.normals(buffers) {
+                iter.map(|x| x.into()).collect()
+            } else {
+                Vec::new()
+            };
+            let tangents = if let Some(iter) = primitive.tangents(buffers) {
+                iter.map(|x| x.into()).collect()
+            } else {
+                Vec::new()
+            };
+            let tex_coords = if let Some(iter) = primitive.tex_coords_f32(0, buffers) {
+                iter.map(|x| x.into()).collect()
+            } else {
+                Vec::new()
+            };
+            let geometry = Geometry {
+                base_shape: Shape {
+                    vertices: vertices,
+                    normals: normals,
+                    tangents: tangents,
+                    tex_coords: tex_coords,
+                },
+                faces: faces,
+                .. Geometry::empty()
+            };
+            let material = self.load_gltf_material(
+                &primitive.material(),
+                buffers,
+                base,
+            );
+            primitives.push(self.mesh(geometry, material));
+        }
+        primitives
+    }
+
+    /// Loads a single `glTF` node.
+    fn load_gltf_node(
+        &mut self,
+        node: &gltf::Node,
+        buffers: &gltf_importer::Buffers,
+        base: &Path,
+        meshes: &mut VecMap<Vec<Mesh>>,
+        instances: &mut Vec<Mesh>,
+    ) -> Group {
+        fn recurse(
```

Comment:
might look cleaner if you unwrap the recursion by using a stack container explicitly

Reply:
I would like to avoid recursion if it's possible. Rust doesn't have TCO, so it can be the source of problems.

Reply:
I'll redesign.